File: Matlab_DIY_functions/demo_SEM_and_mediation.m
%% script to compare mediation and SEM in R

%% working dir
SEM_path = fullfile('P:','boulot','postdoc_CarmenSandi','results','SEM');

%% general parameters
n_samples = 100;
noise_range = 100;

% initial vector
X0 = 1:n_samples;

%% Demo 1: no direct effect
% mediator 1 = transformation of X0 + some noise
noise_X1 = rand(1,n_samples).*noise_range;
path_X0_to_X1 = 1.5;
X1 = 2.3 + path_X0_to_X1.*X0 + noise_X1;

% principal mediator based on transformation of X1
noise_M = rand(1,n_samples).*noise_range;
path_X1_to_M = 3.2;
M = -4.5 + path_X1_to_M.*X1 + noise_M;

% build Y based on transformation of M
noise_Y = rand(1,n_samples).*noise_range;
path_M_to_Y = 3.2;
Y = -4.5 + path_M_to_Y.*M + noise_Y;

%% look at the resulting mediation analysis
[a,b,c,c_prime, pval] = mediation(X1,M,Y);

%% create matrix
mtrx_THE = [X0', X1', M', Y'];
% save data to load in R
save([SEM_path,filesep,'demo_SEM_1.mat'],'mtrx_THE');


%% Demo 2: include both indirect and direct effects on Y
% Direct effect = X1 to Y

% build Y based on transformation of M + X1
noise_Y = rand(1,n_samples).*noise_range;
path_M_to_Y = 3.2;
path_X1_to_Y = 5.4;
Y = -4.5 + path_M_to_Y.*M + path_X1_to_Y.*X1 + noise_Y;
% create matrix
mtrx_THE = [X0', X1', M', Y'];
% save data to load in R
save([SEM_path,filesep,'demo_SEM_2.mat'],'mtrx_THE');

%% Demo 3: include both indirect and direct effects on Y
% Direct effect = X1 to Y + X0 to Y

% build Y based on transformation of M + X1 + X0
noise_Y = rand(1,n_samples).*noise_range;
path_M_to_Y = 3.2;
path_X1_to_Y = 5.4;
path_X0_to_Y = 2.7;
Y = -4.5 + path_M_to_Y.*M + path_X1_to_Y.*X1 + path_X0_to_Y.*X0 + noise_Y;
% create matrix
mtrx_THE = [X0', X1', M', Y'];
% save data to load in R
save([SEM_path,filesep,'demo_SEM_3.mat'],'mtrx_THE');

File: Matlab_DIY_functions/demo_SEM_and_mediation.R
# Demo on SEM to compare with Matlab home-made mediation script

# workind directory
setwd("P:/boulot/postdoc_CarmenSandi/results/SEM")

## load the data
demo1 <- readMat("demo_SEM_1.mat")
mtrx_SEM1 = demo1$mtrx.THE
demo2 <- readMat("demo_SEM_2.mat")
mtrx_SEM2 = demo2$mtrx.THE
demo3 <- readMat("demo_SEM_3.mat")
mtrx_SEM3 = demo3$mtrx.THE

################################################################################
## DEMO 0: compare mediation and SEM
# load variables
X1 = demo1$mtrx.THE[,2]
M = demo1$mtrx.THE[,3]
Y = demo1$mtrx.THE[,4]
#create data frame
df = data.frame(X1, M, Y)

# Specify model
sem_model <- '
M ~ X1
Y ~ M + X1
'

# Estimate our model
fit_model <- sem(sem_model,data=df)
# Summarize the results since it is degree of fred = 0, it is just identified, so fit.measures won't give more
summary(fit_model, fit.measures=TRUE, standardized = TRUE, rsquare=TRUE)
# in case you need to extract p.value very low use the following lines:
# details <- summary(fit_model, fit.measures=TRUE, standardized = TRUE, rsquare=TRUE)
#details$pe
# make a nice display
semPaths(fit_model, "std", edge.label.cex = 1.0, curvePivot = TRUE)


################################################################################
## DEMO 1: only indirect path
# load variables
X0 = demo1$mtrx.THE[,1]
X1 = demo1$mtrx.THE[,2]
M = demo1$mtrx.THE[,3]
Y = demo1$mtrx.THE[,4]

#create data frame
df = data.frame(X0, X1, M, Y)

# Specify model
sem_model <- '
X1 ~ X0
M ~ X1
Y ~ M
'

# Estimate our model
fit_model <- sem(sem_model,data=df)
# Summarize the results since it is degree of fred = 0, it is just identified, so fit.measures won't give more
summary(fit_model, fit.measures=TRUE, standardized = TRUE, rsquare=TRUE)
# in case you need to extract p.value very low use the following lines:
# details <- summary(fit_model, fit.measures=TRUE, standardized = TRUE, rsquare=TRUE)
#details$pe
# make a nice display
semPaths(fit_model, "std", edge.label.cex = 1.0, curvePivot = TRUE)

############################################################################################
## DEMO 2: indirect + direct path (X1=>M)
# load variables
X0 = demo2$mtrx.THE[,1]
X1 = demo2$mtrx.THE[,2]
M = demo2$mtrx.THE[,3]
Y = demo2$mtrx.THE[,4]

#create data frame
df = data.frame(X0, X1, M, Y)

# Specify model
sem_model <- '
X1 ~ X0
M ~ X1
Y ~ M + X1
'

# Estimate our model
fit_model <- sem(sem_model,data=df)
# Summarize the results since it is degree of fred = 0, it is just identified, so fit.measures won't give more
summary(fit_model, fit.measures=TRUE, standardized = TRUE, rsquare=TRUE)
# in case you need to extract p.value very low use the following lines:
# details <- summary(fit_model, fit.measures=TRUE, standardized = TRUE, rsquare=TRUE)
#details$pe
# make a nice display
semPaths(fit_model, "std", edge.label.cex = 1.0, curvePivot = TRUE)

#######################
## DEMO 3: indirect + direct path (X1=>M and X0=>M)
# load variables
X0 = demo3$mtrx.THE[,1]
X1 = demo3$mtrx.THE[,2]
M = demo3$mtrx.THE[,3]
Y = demo3$mtrx.THE[,4]

#create data frame
df = data.frame(X0, X1, M, Y)

# Specify model
sem_model <- '
X1 ~ X0
M ~ X1
Y ~ M + X0 + X1
'

# Estimate our model
fit_model <- sem(sem_model,data=df)
# Summarize the results since it is degree of fred = 0, it is just identified, so fit.measures won't give more
summary(fit_model, fit.measures=TRUE, standardized = TRUE, rsquare=TRUE)
# in case you need to extract p.value very low use the following lines:
# details <- summary(fit_model, fit.measures=TRUE, standardized = TRUE, rsquare=TRUE)
#details$pe
# make a nice display
semPaths(fit_model, "std", edge.label.cex = 1.0, curvePivot = TRUE)